docs: add 'going further' section for custom routing (#19565)
nuxt/nuxt: docs/2.guide/3.going-further/12.custom-router.md

---
title: "Custom router"

> description: "In Nuxt 3, your routing is the defined by the structure your files inside the pages directory. However, since it uses vue-router under the hood, Nuxt offers you severals ways to add custom routes in your project."

Reviewer: ```suggestion
description: "In Nuxt 3, your routing is the defined by the structure your files inside the pages directory. However, since it uses vue-router under the hood, Nuxt offers you several ways to add custom routes in your project."
```

---

## Adding custom routes

In Nuxt 3, your routing is the defined by the structure your files inside the [pages directory](/docs/guide/directory-structure/pages). However, since it uses [vue-router](https://router.vuejs.org/) under the hood, Nuxt offers you several ways to add custom routes in your project.

### Using router config

Using [router options](/docs/2.guide/3.going-further/12.custom-router#router-options),You can optionally override routes using a function that accepts scanned routes and returns customized routes.

> If returning `null` or `undefined`, Nuxt will fallback to the default routes. (useful to modify input array)

Reviewer: ```suggestion

If it returns `null` or `undefined`, Nuxt will fall back to the default routes (useful to modify input array).
```

---
title: "Custom router"

> description: "In Nuxt 3, your routing is the defined by the structure your files inside the pages directory. However, since it uses vue-router under the hood, Nuxt offers you several ways to add custom routes in your project."

Reviewer 2: ```suggestion
description: "In Nuxt 3, your routing is defined by the structure of your files inside the pages directory. However, since it uses vue-router under the hood, Nuxt offers you several ways to add custom routes in your project."
```

---

## Adding custom routes

> In Nuxt 3, your routing is the defined by the structure your files inside the [pages directory](/docs/guide/directory-structure/pages). However, since it uses [vue-router](https://router.vuejs.org/) under the hood, Nuxt offers you several ways to add custom routes in your project.

Reviewer 2: ```suggestion
In Nuxt 3, your routing is defined by the structure of your files inside the [pages directory](/docs/guide/directory-structure/pages). However, since it uses [vue-router](https://router.vuejs.org/) under the hood, Nuxt offers you several ways to add custom routes in your project.
```

### Using router config

> Using [router options](/docs/2.guide/3.going-further/12.custom-router#router-options), you can optionally override routes using a function that accepts scanned routes and returns customized routes.

Reviewer: ```suggestion
Using [router options](/docs/2.guide/3.going-further/12.custom-router#router-options), you can optionally override or extend your routes using a function that accepts the scanned routes and returns customized routes.
```

> Using [router options](/docs/2.guide/3.going-further/12.custom-router#router-options),You can optionally override routes using a function that accepts scanned routes and returns customized routes.

Reviewer 2: ```suggestion
Using [router options](/docs/2.guide/3.going-further/12.custom-router#router-options), you can optionally override routes using a function that accepts scanned routes and returns customized routes.
```

If it returns `null` or `undefined`, Nuxt will fall back to the default routes (useful to modify input array).

```js [app/router.options.ts]
import type { RouterConfig } from '@nuxt/schema'

// https://router.vuejs.org/api/interfaces/routeroptions.html
export default <RouterConfig> {
  routes: (_routes) => [
    {
      name: 'home',
      path: '/',
      component: () => import('~/pages/home.vue')
    }
  ],
}
```

::alert

> Nuxt will not augment any new routes you return from the `routes` function with metadata defined in `definePageMeta` of the component you provide. If you want that to happen, you should probably use the `pages:extend` hook which is [called at build-time](/docs/api/advanced/hooks#nuxt-hooks-build-time).

Reviewer 2: Probably is too ambiguous. Does the user need to use `pages:extend` or not? I think the word "probably" can be removed.

Author: I removed it :)

::

> ### Using pages:extends hook

Reviewer: ```suggestion
### Using the `pages:extend` hook
```

> You can add, change or remove pages from the scanned routes with the `pages:extend` nuxt hook. For example, to prevent creating routes for any `.ts` files:

Reviewer: ```suggestion
You can add, change or remove pages from the scanned routes with the `pages:extend` nuxt hook (or the `extendPages` utility which you can import from `@nuxt/kit`, if you are creating a module). For example, to prevent creating routes for any `.ts` files:
```

### Using `app/router.options`

It is possible to customize [vue-router options](https://router.vuejs.org/api/interfaces/routeroptions.html).

This is the recommended way to specify router options.

```js [app/router.options.ts]
import type { RouterConfig } from '@nuxt/schema'

// https://router.vuejs.org/api/interfaces/routeroptions.html
export default <RouterConfig> {
}
```

### Using `nuxt.config`

**Note:** Only JSON serializable options are configurable:

- `linkActiveClass`
- `linkExactActiveClass`
- `end`
- `sensitive`
- `strict`
- `hashMode`

```js [nuxt.config]
export default defineNuxtConfig({
  router: {
    // https://router.vuejs.org/api/interfaces/routeroptions.html
    options: {}
  }
})
```

### Hash Mode (SPA)

You can enable hash history in SPA mode. In this mode, router uses a hash character (#) before the actual URL that is internally passed. When enabled, the **URL is never sent to the server** and **SSR is not supported**.

```ts [nuxt.config.ts]
export default defineNuxtConfig({
  ssr: false,
  router: {
    options: {
      hashMode: true
    }
  }
})
```

#### Custom History (advanced)

> You can optionally override history mode using a function that accepts base url and returns history mode.
> If returning `null` or `undefined`, Nuxt will fallback to the default history.

Reviewer 2: ```suggestion
You can optionally override history mode using a function that accepts the base URL and returns history mode. If returning `null` or `undefined`, Nuxt will fallback to the default history.
```

Author: I may be more readable the way it is, it was the point of https://github.com/nuxt/nuxt/pull/19565/commits/30fd48f0338f59324ed4d61af93c924f86ca868d

Reviewer: made a small edit - how about now?
